cpp/include/cudf/rolling.hpp

```diff
@@ -275,5 +275,133 @@ std::unique_ptr<column> rolling_window(
   std::unique_ptr<aggregation> const& agg,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief  Applies an offset-based window function (like LEAD/LAG) on specified column.
+ *
+ * Offset-based window functions include LEAD and LAG:
+ *   1. LEAD(N): Returns the row from the input column, at the specified offset past the
+ *      current row. If the offset crosses the grouping boundary or column boundary for
+ *      a given row, a "default" value is returned if available. The "default" value is
+ *      null, by default.
+ *   2. LAG(N): returns the row from the input column at the specified offset preceding
+ *      the current row. If the offset crosses the grouping boundary or column boundary for
+ *      a given row, a "default" value is returned if available. The "default" value is
+ *      null, by default.
+ *
+ * E.g. Consider an input column with the following values and grouping:
+ *      [10, 11, 12, 13,   20, 21, 22, 23]
+ *      <------G1----->   <------G2------>
+ *
+ * LEAD(input_col, 1) yields:
+ *      [11, 12, 13, null,  21, 22, 23, null]
+ *
+ * LAG(input_col, 2) yields:
+ *      [null, null, 11, 12, null, null, 21, 22]
+ *
+ * LEAD(input_col, 1, 99) (where 99 indicates the default) yields:
+ *      [11, 12, 13, 99,  21, 22, 23, 99]
+ *
+ * @param[in] group_keys The (pre-sorted) grouping columns
+ * @param[in] input The input column (to be aggregated)
+ * @param[in] aggr The aggregation (e.g. LEAD/LAG) to be applied
+ * @param[in] row_offset The offset from which input row is to be returned
+ * @param[in] default_outputs A column of default outputs, to be returned
+ *                            if the offset out of bounds. Must have as many rows
+ *                            as the input column, or be empty.
+ *
+ * @returns  A nullable output column containing the results of the window operation.
+ */
+std::unique_ptr<column> offset_rolling_window(
```

Comment:
I wouldn't think this would need it's on rolling_window API. I would hope that any aggregation specific info can be put into the `aggregation` object.

Reply:
That is where I started, actually: putting this in `grouped_rolling_window()`. 

Lead/Lag are a little different from the other aggregations:
  1. They work with a single offset value, interpreted based on direction.
  2. They support default values where other operators don't. 
  3. The window width calculation ended up differing from what we have in the `grouped_` case. 

In the end, there was `Lead`/`Lag`-specific error-handling that obfuscated `grouped_rolling_window()`. Moving it out to another function simplified everything. 

An earlier version of this function was an overload of `grouped_rolling_window`. It was renamed to avoid implying support for all window operators, and to reorder the parameters more appropriately for `Lead`/`Lag`.

Reply:
> I would hope that any aggregation specific info can be put into the aggregation object.

Oh, I think I see what you mean now. Let me try cleaning up the API, and putting the `row_offset` into `LEAD`/`LAG`.

Reply:
I have been trying to evaluate whether it also makes sense to move the default outputs into the aggregation operator payload, as a scalar. I don't think that would be viable. The caller might use an expression as a default, thus yielding a default column.
I'll try moving the offset into the aggregation operator itself.

Reply:
I took a crack at writing a `LEAD`/`LAG` aggregation that stores a `row_offset`. I'm afraid it doesn't look possible to pass aggregation objects with extra metadata into the `process_rolling_window()` device function. 
The aggregation operation in `rolling_window` seems to be identified by a stateless `agg_op` template parameter. :/ This doesn't look doable.

Reply:
> I would hope that any aggregation specific info can be put into the aggregation object.

@jrhemstad, I have tried to accommodate your suggestion over at #6277. If that version seems more palatable, I will close this PR in favour of that approach.